src/Bounds.cpp

```diff
@@ -1039,11 +1039,29 @@ class Bounds : public IRVisitor {
                     if (op->is_intrinsic(Call::shift_left)) {
                         if (t.is_int() && t.bits() >= 32) {
                             // Overflow is UB
-                            if (a_interval.has_lower_bound() && b_interval.has_lower_bound() && can_prove(b_interval.min >= 0 && b_interval.min < t.bits())) {
+                            if (a_interval.has_lower_bound() &&
+                                b_interval.has_lower_bound() &&
+                                can_prove(b_interval.min >= 0 &&
+                                          b_interval.min < t.bits())) {
                                 interval.min = a_interval.min << b_interval.min;
+                            } else if (a_interval.has_lower_bound() &&
+                                       b_interval.has_lower_bound() &&
+                                       !b_interval.min.type().is_uint() &&
+                                       can_prove(b_interval.min < 0 &&
+                                                 cast<int>(b_interval.min) > -t.bits())) {
```

Comment:
You don't need the cast anymore, and it would be better not to have it if possible.

Reply:
Good point. Removed.